backend/conversationtelemetry.py

```diff
@@ -0,0 +1,69 @@
+    It allows tracking and storing conversation details such as user input, model used, and response generated.
+
+    Attributes:
+        cosmosdb_endpoint (str): The endpoint URL of the Cosmos DB.
+        credential (any): The credential used to authenticate with the Cosmos DB.
+        database_name (str): The name of the database in the Cosmos DB.
+        container_name (str): The name of the container in the database.
+        cosmosdb_client (CosmosClient): The client object for interacting with the Cosmos DB.
+        database_client (DatabaseClient): The client object for interacting with the database.
+        container_client (ContainerClient): The client object for interacting with the container.
+    """
+
+    def __init__(self, cosmosdb_endpoint: str, credential: any, database_name: str, container_name: str):
+        self.cosmosdb_endpoint = cosmosdb_endpoint
+        self.credential = credential
+        self.database_name = database_name
+        self.container_name = container_name
+        self.cosmosdb_client = CosmosClient(self.cosmosdb_endpoint, credential=credential)
+        self.database_client = self.cosmosdb_client.get_database_client(database_name)
+        self.container_client = self.database_client.get_container_client(container_name)
+
+    def create_conversation_item(self, request_body, aoai_service, model, temperature):
+
+        id = str(uuid.uuid4())
+        if len(request_body["messages"]) > 0 and request_body["messages"][-1]['role'] == "user":
+            conversation_item = {
+                'aoai_service': aoai_service,
+                'model': model,
+                'temperature': temperature,
+                'timestamp': datetime.now().isoformat(),
+                'user_input': request_body["messages"][-1],
+                'user_id': str(uuid.uuid4()),
+                'conversation_id': request_body["messages"][0]["id"],
+                'id': id
+            }
+            self.container_client.upsert_item(conversation_item)
+        return id
+
+    def update_conversation_item(self, result, conversation_item_id, error: Exception = None):
+
+        query = f"SELECT * FROM c WHERE c.id = '{conversation_item_id}'"
+        items = list(self.container_client.query_items(query, enable_cross_partition_query=True))
+        conversation_item = items[0] if len(items) else None
+        role = result["choices"][0]["messages"][0]["role"]
+        if conversation_item is None:
+            return
+        else:
+            conversation_item["response_timestamp"] = datetime.now().isoformat()
+            if error is not None:
+                conversation_item["error"] = str(error)
+                conversation_item["traceback"] = traceback.format_exc()
+            else:
+                if role == "tool":
+                    conversation_item["tool"] = result["choices"][0]["messages"][0]["content"]
```

Comment:
It looks like the document content and intent are stored in the custom cosmos telemetry as a JSON string. 

![image](https://github.com/ciesko/msrchat/assets/95880385/cdde9ced-c333-4e0a-b9c5-33aacefee768)

If possible to be done cleanly and safely, this could be improved by sending it as a Python dict instead (as in the RCT cosmos logs).

![image](https://github.com/ciesko/msrchat/assets/95880385/f7c9cab2-76f4-47be-80af-9db1e3385179)
_Not suggesting changing field names or matching the object property names, just showing what the RCT one looks like for comparison_